src/compressor/QatAccel.cc

```diff
@@ -25,16 +42,53 @@ void QzSessionDeleter::operator() (struct QzSession_S *session) {
 /* Estimate data expansion after decompression */
 static const unsigned int expansion_ratio[] = {5, 20, 50, 100, 200};
 
-QatAccel::QatAccel() {
-  session.reset(new struct QzSession_S);
-  memset(session.get(), 0, sizeof(struct QzSession_S));
-}
+// put the session back to the session pool in a RAII manner
+struct cached_session_t {
+  cached_session_t(QatAccel* accel)
+    : accel{accel} {
+    std::unique_lock lock{accel->mutex};
+    accel->cond.wait(lock, [accel] { return !accel->sessions.empty(); });
```

Comment:
due to rgw's use of coroutines for request handling, we've been trying to remove all blocking waits like this from our request path. with coroutines, it's much better if we can suspend the coroutine and use the thread to work on another request, instead of blocking the entire thread here

is it really necessary to block on empty here? maybe once we exceed `qat_compressor_session_number`, we start allocating new sessions for each request instead of waiting for one to finish? then the destructor would only push_back() if `sessions.size() < qat_compressor_session_number`, and free the session otherwise

Reply:
> QAT can't free session during work

do we understand why? when is it safe to free them?

Reply:
I am still looking for the reason.

Reply:
Hi, @cbodley , I found the reason.  In QATzip, `qzClose()` will close the connection with QAT, which leads all session to unavailable status. So it lead `qzCompress()` to crash after calling `qzClose()`.  Now I place `qzClose()` on `~QatAccel`, it works well.